text/0000-oracles.md

```diff
@@ -0,0 +1,136 @@
+| Status  | Stakeholders | Outcome | Due date |   Owner   |
+| :-----: | :----------: | :-----: | :------: | :-------: |
+| Started |              |         |          | @arjentix |
+
+
+
+## Background
+
+"Oracle" is a common name for different technologies which basically do one thing – bring off-chain data on chain. The possibilities are almost limitless:
+
+1. Real wold currencies
+2. Random numbers generation
+3. Timestamp retrieving
+4. Events in other blockchains
+5. Election results
+6. Geo-data i.e. to track delivery chains
+7. Complex calculations
+8. Weather data
+9. Statistical data
+10. The central bank's key rate
+
+Obviously integrating Oracles into Iroha 2 based blockchains will expand the horizon on possibilities for our users. By providing a clear way to integrate your WASM Trigger with Oracles we encourage people to use Iroha 2 in really complex systems and bring their ideas to life.
+
+## Problem
+
+We need to add Oracles support into Iroha 2. Oracles should be a flexible part of a blockchain and should not be hardcoded anywhere on our side. There should be a clear mechanism to interact with them and respective documentation.
+
+### Example scenarios
+
+ Scenario 1:
+
+1. End-user interacts with a trigger to buy some asset which is tied to a real world, i.e. gold, for some other asset A;
+2. Trigger calls Oracle 1 to get gold price in dollars, and then calls Oracle 2 to get asset A price in dollars;
+3. Trigger computes the result and transfers money to the buyer.
+
+Scenario 2:
+
+1. Weather sensors have registered a hurricane
+2. Trigger has been activated to transfer insurance benefits to the clients
+
+ Scenario 3:
+
+1. End-user interacts with a trigger, requesting some permission token to be given to their account. This permission can be given only to account which corresponds to University "X" students.
+2. Trigger calls Oracle of University "X" to ensure that account corresponds to a student
+3. On success trigger sends permission token to the account
+
+## Solution
+
+Proposed solution is based on already built parts of Iroha and does not require much efforts to implement.
+It consists of the two parts:
+
+1. `EmitOracleRequest` Iroha Special Instruction;
+2. `OracleRequest` event type.
+
+`EmitOracleRequest`  – is a new ISI that does one thing – emits `OracleRequest` with provided payload.
+
+`OracleRequest`  – special event which directly says that blockchain asks for off-chain data from some Oracle. Such events will be transmitted in event pipeline like all other Iroha 2 events.
```

Comment:
> like all other events

Triggers might be set up to listen for Data, Pipeline and Time events. Will they support `OracleRequest` event?

Reply:
In the current design they can

Reply:
So, it makes a step towards easier implementation of "Turing-complete" ISA: instructions might register triggers and emit events which will trigger those triggers. Or I miss something?